### DISCUSSION TITLE
Would it be possible to somehow identify Firefox windows despite changing window titles?

### DISCUSSION THREAD

**Discussion by EkanshdeepGupta** [nlpsuge/gnome-shell-extension-another-window-session-manager#61] [Ideas]
First of all, I love the extension, huge fan. This was such a pain point for me that I had half a mind of writing something like this myself, until I stumbled upon your repository. It is adding so much value to my GNOME experience.

The only feature I would want (other than the ability to bind keyboard shortcuts), is for the extension to somehow be able to identify the few different Firefox windows I have, and place them appropriately as well. I know at present it recognizes windows using the application name and window title. The only problem is I keep switching to different tabs, so the window title does not remain constant.

Would it be possible to have some kind of way to identify different Firefox windows (I know, what does it mean for a Firefox window to be same as another Firefox window in previous session? That is why I am starting a discussion to see if you have any ideas). Maybe I can use some Firefox extensions to help this extension identify windows? What do you think?

My current workflow looks like as follows: I have a few different projects. I usually have all windows related to one project in a different virtual desktop to keep things organized. As part of each project, I have a different Firefox window which keeps tabs related to each project disjoint and organized. While Another Window Session Manager is great at putting all the other windows in their right places, it (understandably) struggles with Firefox windows because their window titles keep changing. As a result, I have to resort to manually looking and placing all the Firefox windows to their correct desktops quite often (say whenever I plug/unplug a display or reboot). I was looking for just any kind of solution which would make this kind of (not too atypical) situation better.

**Comment by nlpsuge** [comment]
> Would it be possible to have some kind of way to identify different Firefox windows (I know, what does it mean for a Firefox window to be same as another Firefox window in previous session? That is why I am starting a discussion to see if you have any ideas). Maybe I can use some Firefox extensions to help this extension identify windows? What do you think?

It's possible to add a unique string to all the titles of a window. This modification can be saved by Firefox and this extension and is used to identify the window. I know at least two Firefox extensions can do this job, https://addons.mozilla.org/en-US/firefox/addon/crappy-firetitle/ and https://addons.mozilla.org/en-US/firefox/addon/window-titler/.  For example, if the original title is `Gmail — Mozilla Firefox`, the modified title will be like `Work for something -- Gmail — Mozilla Firefox`. If the separator is `--`, the window unique id is `Work for something`.

The user experience might not be very good. Because it requires users to install a browser extension and they have to setup the settings of that extension and the settings of this gnome-shell extension. 

But it's probably better than nothing.

I'm thinking to support `idea`, `vs code` etc.

I have not found a Chrome extension to do the same thing like what https://addons.mozilla.org/en-US/firefox/addon/crappy-firetitle/ does.
 
I'm still not sure if there are any other easier ways to identify the windows.

**Comment by EkanshdeepGupta** [comment]
That's perfect. I had an additional question, is there any way to use regex in the sessions json file? I want to write something like `"window-title": "[Work]*"`, in other words only match a prefix of the string.

**Reply by nlpsuge** [reply]
> is there any way to use regex in the sessions json file? I want to write something like `"window-title": "[Work]*"`

You can do this by modifying the window title using a Firefox extension, this extension will save it to the session json file.

> in other words only match a prefix of the string

I don't know why do you want to do that, what's the user case?

**Comment by EkanshdeepGupta** [comment]
Okay, so I tried searching, but I could not find any Firefox extension which completely replaces the Firefox title. The two extensions you mentioned are only able to add a prefix to the window title. That is why I was looking for a way for Another Window Session Manager to match with prefixes.

If you find a way to completely overwrite a Firefox window title, I would be delighted because that would solve my issue.
